File: 4 - Class vs. Instance/Solution.fs
open System

type Person(initialAge) = 
    let mutable age = 0
    
    do 
        if initialAge > 0 then age <- initialAge
        else Console.WriteLine("Age is not valid, setting age to 0.")
    
    member val Age = age with get, set
    
    member this.amIOld() = 
        if this.Age < 13 then Console.WriteLine("You are young.")
        elif this.Age < 18 then Console.WriteLine("You are a teenager.")
        else Console.WriteLine("You are old.")
    
    member this.yearPasses() = this.Age <- this.Age + 1

[<EntryPoint>]
let main argv = 
    let n = Console.ReadLine() |> int
    for i = 0 to n do
        let age = Console.ReadLine() |> int
        let person = Person(age)
        person.amIOld()
        for j = 0 to 3 do
            person.yearPasses()
        person.amIOld()
        Console.WriteLine()
    0

File: 4 - Class vs. Instance/Solution.go
package main

import "fmt"

type person struct {
	age int
}

func (p person) NewPerson(initialAge int) person {
	if initialAge < 0 {
		fmt.Println("Age is not valid, setting age to 0.")
		p.age = 0
	}

	return p
}

func (p person) amIOld() {
	if p.age < 13 {
		fmt.Println("You are young.")
	} else if p.age >= 13 && p.age < 18 {
		fmt.Println("You are a teenager.")
	} else {
		fmt.Println("You are old.")
	}
}

func (p person) yearPasses() person {
	p.age++
	return p
}

func main() {
	var T, age int

	fmt.Scan(&T)

	for i := 0; i < T; i++ {
		fmt.Scan(&age)
		p := person{age: age}
		p = p.NewPerson(age)
		p.amIOld()
		for j := 0; j < 3; j++ {
			p = p.yearPasses()
		}
		p.amIOld()
		fmt.Println()
	}
}
